src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +174,105 @@ private sealed class RunningInstance
+                    if(_watcherRunLoop == IntPtr.Zero)
+                    {
+                        lock (lockObject)
+                        {
+                            if (_watcherRunLoop == IntPtr.Zero)
+                            {
+                                var runLoopStarted = new ManualResetEventSlim();
+                                new Thread(WatchForFileSystemEventsThreadStart) { IsBackground = true }.Start(new object[] { runLoopStarted, eventStream });
+                                runLoopStarted.Wait();
+
+                                Interlocked.Increment(ref scheduledStreamsCount);
+                                return;
+                            }
+                        }
+                    }
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                    Interlocked.Increment(ref scheduledStreamsCount);
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (_watcherRunLoop != IntPtr.Zero)
+                    {
+                        // Always unschedule the RunLoop before cleaning up
+                        Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                        Interlocked.Decrement(ref scheduledStreamsCount);
+
+                        if (scheduledStreamsCount == 0)
+                        {
+                            // Stop the FS event message pump
+                            Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
                 {
-                    if (!_stopping && _watcherRunLoop != IntPtr.Zero)
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    _watcherRunLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    Debug.Assert(_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(_watcherRunLoop);
+                    Debug.Assert(retainResult == _watcherRunLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
```

Comment:
Why is the first call / the one spinning up the thread different from all of the others? i.e. why does this call need to be part of WatchForFilesystemEventsThreadStart rather than just being in ScheduleEventStream?

Reply:
RunLoop won't run without any scheduled event stream.